io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -129,7 +131,7 @@ data TimerCompletionInfo s =
      | TimerThreadDelay !ThreadId !TimeoutId
      -- ^ `threadDelay` timer run by `ThreadId` which was assigned the given
      -- `TimeoutId` (only used to report in a trace).
-     | TimerTimeout !ThreadId !TimeoutId !(STRef s IsLocked)
+     | TimerTimeout !ThreadId !TimeoutId !(TMVar (IOSim s) ThreadId)
```

Comment:
Why changing to a TMVar?

Reply:
The primary reason was that I wanted to write thread actions using `IOSim` not `SimA`.  It turned out that I had to use the `execReadTVar` but in a limited way.

We actually should use an `MVar`, but it doesn't matter in our use case.  We have only one thread that runs `takeMVar`, so we don't need fairness.  We have two threads that try to write to it, but it's guaranteed that one of them receives the `TMVar` empty, so fairness is not a problem here either.

Reply:
Okay got it! I don't mind the change!